context_response_test.go

```diff
@@ -701,3 +702,86 @@ func TestResponse_Status(t *testing.T) {
 		})
 	}
 }
+
+func TestResponse_Stream(t *testing.T) {
+	var (
+		err        error
+		fiber      *Route
+		req        *http.Request
+		mockConfig *configmocks.Config
+	)
+
+	beforeEach := func() {
+		mockConfig = &configmocks.Config{}
+		mockConfig.On("GetBool", "http.drivers.fiber.prefork", false).Return(false).Once()
+		mockConfig.On("GetInt", "http.drivers.fiber.body_limit", 4096).Return(4096).Once()
+		mockConfig.On("GetInt", "http.drivers.fiber.header_limit", 4096).Return(4096).Once()
+		ConfigFacade = mockConfig
+	}
+
+	tests := []struct {
+		name   string
+		method string
+		url    string
+		setup  func(method, url string) error
+		output []string
+	}{
+		{
+			name:   "Stream",
+			method: "GET",
+			url:    "/stream",
+			setup: func(method, url string) error {
+				fiber.Get("/stream", func(ctx contractshttp.Context) contractshttp.Response {
+					return ctx.Response().Status(http.StatusCreated).Stream(func(w contractshttp.StreamWriter) error {
+						b := []string{"a", "b", "c", "c", "c", "c", "c"}
+						for _, a := range b {
+							if _, err := w.Write([]byte(a + "\n")); err != nil {
+								return err
+							}
+
+							if err := w.Flush(); err != nil {
+								return err
+							}
+						}
+
+						return nil
+					})
+				})
+
+				var err error
+				req, err = http.NewRequest(method, url, nil)
+				if err != nil {
+					return err
+				}
+
+				return nil
+			},
+			output: []string{"a", "b", "c", "c", "c", "c", "c"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			beforeEach()
+			fiber, err = NewRoute(mockConfig, nil)
+			assert.Nil(t, err)
+
+			err := test.setup(test.method, test.url)
+			assert.Nil(t, err)
+
+			resp, err := fiber.Test(req)
+			assert.NoError(t, err)
+
+			scanner := bufio.NewScanner(resp.Body)
+			var output []string
+			for scanner.Scan() {
+				output = append(output, scanner.Text())
+			}
```

Comment:
I doubt that this way can't test the stream accurately. I asked GPT, could you try the code below?

```
package main

import (
    "time"

    "github.com/gofiber/fiber/v2"
)

func streamData(c *fiber.Ctx) error {
    c.Set(fiber.HeaderContentType, fiber.MIMETextHTMLCharsetUTF8)

    streamErr := c.Context().SetBodyStreamWriter(func(w *fiber.StreamWriter) error {
        for i := 0; i < 10; i++ {
            _, err := w.Write([]byte("Streaming data " + string(i+48) + "..."))
            if err != nil {
                return err
            }
            time.Sleep(100 * time.Millisecond) // Simulate delay between chunks
        }
        return nil
    })

    if streamErr != nil {
        return streamErr
    }

    return nil
}

func main() {
    app := fiber.New()

    app.Get("/stream", streamData)

    app.Listen(":3000")
}
```

```
package main

import (
    "bufio"
    "net/http/httptest"
    "strings"
    "testing"
    "time"

    "github.com/gofiber/fiber/v2"
    "github.com/stretchr/testify/assert"
)

func TestStreamData(t *testing.T) {
    app := fiber.New()

    app.Get("/stream", streamData)

    req := httptest.NewRequest("GET", "/stream", nil)
    resp, err := app.Test(req, -1) // -1 means read until we get the full response or the connection is closed

    assert.NoError(t, err)
    assert.Equal(t, 200, resp.StatusCode)
    assert.Equal(t, resp.Header.Get("Content-Type"), fiber.MIMETextHTMLCharsetUTF8)

    expectedChunks := []string{
        "Streaming data 0...",
        "Streaming data 1...",
        "Streaming data 2...",
        "Streaming data 3...",
        "Streaming data 4...",
        "Streaming data 5...",
        "Streaming data 6...",
        "Streaming data 7...",
        "Streaming data 8...",
        "Streaming data 9...",
    }

    reader := bufio.NewReader(resp.Body)
    for i, expected := range expectedChunks {
        chunk, err := reader.ReadString(' ')
        if err != nil {
            t.Fatalf("Failed to read chunk: %v", err)
        }
        chunk += ".." // Append rest of the string
        rest, _ := reader.ReadString('.')
        chunk += rest

        assert.Equal(t, expected, strings.TrimSuffix(chunk, "\n"), "Mismatch in chunk %d", i+1)
        // To simulate reading it with an interval as in real-time
        time.Sleep(100 * time.Millisecond)
    }
}
```

Reply:
Both are similar

Reply:
I tested them, and yes, both of them get the final response(a,b,c), then scan the result, this way can't test the output data verbatim. It's not the perfect solution, but I think it's good enough for us now. Great job.